SYCL/DeviceLib/math_test_marray_vec.cpp

```diff
@@ -0,0 +1,279 @@
+                        T{+0, +0, +0, +0}, 0);
+  math_test_3_fma(deviceQueue, T{-1, -1, -1, -1}, T{2, 2, 2, 2}, T{2, 2, 2, 2},
+                  T{1, 1, 1, 1}, 5);
+  math_test_3_mad(deviceQueue, T{-1, -1, -1, -1}, T{2, 2, 2, 2}, T{2, 2, 2, 2},
+                  T{1, 1, 1, 1}, 5);
+  math_test_3_mix(deviceQueue, T{-1, -1, -1, -1}, T{3, 3, 3, 3}, T{5, 5, 5, 5},
+                  T{0.5, 0.5, 0.5, 0.5}, 4);
+}
+
+template <typename T> void math_tests_3(queue &deviceQueue) {
+  math_test_tan(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_tanh(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_cos(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 1);
+  math_test_sin(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_cosh(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 1);
+  math_test_sinh(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_acos(deviceQueue, T{-1, -1, -1}, T{1, 1, 1}, 0);
+  math_test_acosh(deviceQueue, T{-1, -1, -1}, T{1, 1, 1}, 0);
+  math_test_asin(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_asinh(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_cbrt(deviceQueue, T{-1, -1, -1}, T{1, 1, 1}, 1);
+  math_test_atan(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_atanh(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_exp(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 1);
+  math_test_exp2(deviceQueue, T{-1, -1, -1}, T{2, 2, 2}, 4);
+  math_test_exp10(deviceQueue, T{-1, -1, -1}, T{2, 2, 2}, 100);
+  math_test_expm1(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_ceil(deviceQueue, T{-1, -1, -1}, T{0.6, 0.6, 0.6}, 1);
+  math_test_tgamma(deviceQueue, T{-1, -1, -1}, T{1, 1, 1}, 1);
+  math_test_lgamma(deviceQueue, T{-1, -1, -1}, T{1, 1, 1}, 0);
+  math_test_erf(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_erfc(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 1);
+  math_test_log(deviceQueue, T{-1, -1, -1}, T{1, 1, 1}, 0);
+  math_test_log2(deviceQueue, T{-1, -1, -1}, T{4, 4, 4}, 2);
+  math_test_log10(deviceQueue, T{-1, -1, -1}, T{100, 100, 100}, 2);
+  math_test_log1p(deviceQueue, T{-1, -1, -1}, T{0, 0, 0}, 0);
+  math_test_logb(deviceQueue, T{-1, -1, -1}, T{1.1, 1.1, 1.1}, 0);
+  math_test_sqrt(deviceQueue, T{-1, -1, -1}, T{4, 4, 4}, 2);
+  math_test_rsqrt(deviceQueue, T{-1, -1, -1}, T{0.25, 0.25, 0.25}, 2);
+  math_test_rint(deviceQueue, T{-1, -1, -1}, T{2.9, 2.9, 2.9}, 3);
+  math_test_round(deviceQueue, T{-1, -1, -1}, T{0.5, 0.5, 0.5}, 1);
+  math_test_trunc(deviceQueue, T{-1, -1, -1}, T{1.9, 1.9, 1.9}, 1);
+}
+
+int main() {
+  queue deviceQueue;
+  math_tests_4<float4>(deviceQueue);
+  math_tests_4<double4>(deviceQueue);
+  math_tests_4<marray<float, 4>>(deviceQueue);
+  math_tests_4<marray<double, 4>>(deviceQueue);
+
+  math_tests_3<float3>(deviceQueue);
+  math_tests_3<double3>(deviceQueue);
+  math_tests_3<marray<float, 3>>(deviceQueue);
+  math_tests_3<marray<double, 3>>(deviceQueue);
+
+  if (deviceQueue.get_device().has(sycl::aspect::fp16)) {
```

Comment:
Having these together with kernels that do not require the aspect can cause certain backends to fail if they don't support fp16 as they try to build them together. One option would be to add `-fsycl-device-code-split=per_kernel` to the build command or to split the test into two tests; one that should always work and one that only runs kernels if the aspect is supported by the device.

Of the two options mentioned, the latter - though a little more complicated - comes with the benefit that we could be able to skip it through LIT in the future through feature checks rather than skipping at runtime making the test look like it passes without actually running anything meaningful.

This likely also applies to SYCL/DeviceLib/built-ins/ext_native_math.cpp.

Reply:
Thanks, I went with the second option and removed `-fsycl-device-code-split=per_kernel` from the non fp16 tests compilations.

Reply:
Ah, I completely missed the split. Thanks for splitting it up though. This should be better for both performance and for the possible future skipping of tests without aspects.